### PR TITLE
feat(adyen provider): Add adyen payment provider

### PR DESCRIPTION
**Pull request by ivannovosad**
# Roadmap Task

👉  https://getlago.canny.io/feature-requests/p/support-integration-with-adyen

## Context

Adyen provides payment apis for card and other payments.

## Description

Add Adyen as a payment provider supported natively

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -40,6 +40,7 @@ gem 'graphql-pagination'
 # Payment processing
 gem 'gocardless_pro', '~> 2.34'
 gem 'stripe'
+gem 'adyen-ruby-api-library'
```
Comment (jdenquin) [inline Gemfile:43]: Bundler/OrderedGems: Gems should be sorted in an alphabetical order within their section of the Gemfile. Gem `adyen-ruby-api-library` should appear before `stripe`.

File: app/graphql/mutations/payment_providers/adyen.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    class Adyen < BaseMutation
+      include AuthenticableApiUser
+      include RequiredOrganization
+
+      graphql_name 'AddAdyenPaymentProvider'
+      description 'Add or update Adyen payment provider'
+
+      argument :api_key, String, required: true
+      argument :merchant_account, String, required: true
+      argument :live_prefix, String, required: false
```
Comment (jdenquin) [inline app/graphql/mutations/payment_providers/adyen.rb:14]: GraphQL/ExtractInputType: Consider moving arguments to a new input type

Comment (jdenquin) [inline app/graphql/mutations/payment_providers/adyen.rb:14]: GraphQL/OrderedArguments: Arguments should be sorted in an alphabetical order within their section. Field `live_prefix` should appear before `merchant_account`.

File: app/graphql/mutations/payment_providers/adyen.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      argument :hmac_key, String, required: false
```
Comment (jdenquin) [inline app/graphql/mutations/payment_providers/adyen.rb:15]: GraphQL/ExtractInputType: Consider moving arguments to a new input type

Comment (jdenquin) [inline app/graphql/mutations/payment_providers/adyen.rb:15]: GraphQL/OrderedArguments: Arguments should be sorted in an alphabetical order within their section. Field `hmac_key` should appear before `live_prefix`.

File: app/graphql/types/mutation_type.rb [hunk 4]
```diff
@@ -46,6 +46,7 @@ class MutationType < Types::BaseObject
     field :destroy_payment_provider, mutation: Mutations::PaymentProviders::Destroy
 
     field :add_gocardless_payment_provider, mutation: Mutations::PaymentProviders::Gocardless
+    field :add_adyen_payment_provider, mutation: Mutations::PaymentProviders::Adyen
```
Comment (jdenquin) [inline app/graphql/types/mutation_type.rb:49]: GraphQL/OrderedFields: Fields should be sorted in an alphabetical order within their section. Field `add_adyen_payment_provider` should appear before `add_gocardless_payment_provider`.

File: app/graphql/types/organization_type.rb [hunk 5]
```diff
@@ -33,6 +33,7 @@ class OrganizationType < Types::BaseObject
 
     field :gocardless_payment_provider, Types::PaymentProviders::Gocardless, null: true
     field :stripe_payment_provider, Types::PaymentProviders::Stripe, null: true
+    field :adyen_payment_provider, Types::PaymentProviders::Adyen, null: true
```
Comment (jdenquin) [inline app/graphql/types/organization_type.rb:36]: GraphQL/OrderedFields: Fields should be sorted in an alphabetical order within their section. Field `adyen_payment_provider` should appear before `stripe_payment_provider`.

File: app/graphql/types/payment_providers/adyen.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+module Types
+  module PaymentProviders
+    class Adyen < Types::BaseObject
+      graphql_name 'AdyenProvider'
+
+      field :id, ID, null: false
+      field :api_key, String, null: false
```
Comment (jdenquin) [inline app/graphql/types/payment_providers/adyen.rb:9]: GraphQL/OrderedFields: Fields should be sorted in an alphabetical order within their section. Field `api_key` should appear before `id`.

File: app/graphql/types/payment_providers/adyen.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      field :merchant_account, String, null: false
+      field :live_prefix, String, null: true
```
Comment (jdenquin) [inline app/graphql/types/payment_providers/adyen.rb:11]: GraphQL/OrderedFields: Fields should be sorted in an alphabetical order within their section. Field `live_prefix` should appear before `merchant_account`.

File: app/graphql/types/payment_providers/adyen.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+      field :hmac_key, String, null: true
```
Comment (jdenquin) [inline app/graphql/types/payment_providers/adyen.rb:12]: GraphQL/OrderedFields: Fields should be sorted in an alphabetical order within their section. Field `hmac_key` should appear before `live_prefix`.

File: app/services/payment_providers/adyen_service.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+module PaymentProviders
+  class AdyenService < BaseService
+    def create_or_update(**args)
+      adyen_provider = PaymentProviders::AdyenProvider.find_or_initialize_by(
+        organization_id: args[:organization].id
```
Comment (jdenquin) [inline app/services/payment_providers/adyen_service.rb:7]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: spec/factories/payment_providers.rb [hunk 10]
```diff
@@ -22,4 +22,20 @@
       { access_token: SecureRandom.uuid }.to_json
     end
   end
+
+  factory :adyen_provider, class: 'PaymentProviders::AdyenProvider' do
+    organization
+    type { 'PaymentProviders::AdyenProvider' }
+
+    secrets do
+      { api_key:, merchant_account:, live_prefix:, hmac_key:}.to_json
```
Comment (jdenquin) [inline spec/factories/payment_providers.rb:31]: Layout/SpaceInsideHashLiteralBraces: Space inside } missing. (https://rubystyle.guide#spaces-braces)

File: spec/models/payment_providers/adyen_provider_spec.rb [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+  describe '#api_key' do
+    subject { provider.api_key }
+
+    let(:api_key) { SecureRandom.uuid }
+
+    before { provider.api_key = api_key }
+
+    it 'returns the api key' do
+      expect(subject).to eq api_key
```
Comment (jdenquin) [inline spec/models/payment_providers/adyen_provider_spec.rb:16]: RSpec/NamedSubject: Name your test subject if you need to reference it explicitly. (https://rspec.rubystyle.guide/#use-subject, https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/NamedSubject)

File: spec/models/payment_providers/adyen_provider_spec.rb [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    end
+  end
+
+  describe '#merchant_account' do
+    subject { provider.merchant_account }
+
+    let(:merchant_account) { "TestECOM" }
```
Comment (jdenquin) [inline spec/models/payment_providers/adyen_provider_spec.rb:23]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

File: spec/models/payment_providers/adyen_provider_spec.rb [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+    before { provider.merchant_account = merchant_account }
+
+    it 'returns the merchant account' do
+      expect(subject).to eq merchant_account
```
Comment (jdenquin) [inline spec/models/payment_providers/adyen_provider_spec.rb:28]: RSpec/NamedSubject: Name your test subject if you need to reference it explicitly. (https://rspec.rubystyle.guide/#use-subject, https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/NamedSubject)

File: spec/models/payment_providers/adyen_provider_spec.rb [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+  describe '#live_prefix' do
+    subject { provider.live_prefix }
+
+    let(:live_prefix) { Faker::Internet.domain_word }
+
+    before { provider.live_prefix = live_prefix }
+
+    it 'returns the live prefix' do
+      expect(subject).to eq live_prefix
```
Comment (jdenquin) [inline spec/models/payment_providers/adyen_provider_spec.rb:40]: RSpec/NamedSubject: Name your test subject if you need to reference it explicitly. (https://rspec.rubystyle.guide/#use-subject, https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/NamedSubject)

File: spec/models/payment_providers/adyen_provider_spec.rb [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+  describe '#hmac_key' do
+    subject { provider.hmac_key }
+
+    let(:hmac_key) { SecureRandom.uuid }
+
+    before { provider.hmac_key = hmac_key }
+
+    it 'returns the hmac key' do
+      expect(subject).to eq hmac_key
```
Comment (jdenquin) [inline spec/models/payment_providers/adyen_provider_spec.rb:52]: RSpec/NamedSubject: Name your test subject if you need to reference it explicitly. (https://rspec.rubystyle.guide/#use-subject, https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/NamedSubject)

File: spec/services/payment_providers/adyen_service_spec.rb [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    end
+
+    context 'with validation error' do
+      let(:token) { nil }
+
+      it 'returns an error result' do
+        result = adyen_service.create_or_update(
+          organization:,
+          api_key: nil,
+          merchant_account: nil
```
Comment (jdenquin) [inline spec/services/payment_providers/adyen_service_spec.rb:49]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: app/graphql/mutations/payment_providers/adyen.rb [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+      include AuthenticableApiUser
+      include RequiredOrganization
+
+      graphql_name 'AddAdyenPaymentProvider'
+      description 'Add or update Adyen payment provider'
+
+      argument :api_key, String, required: true
+      argument :hmac_key, String, required: false
+      argument :live_prefix, String, required: false
+      argument :merchant_account, String, required: true
```
Comment (jdenquin) [inline app/graphql/mutations/payment_providers/adyen.rb:15]: GraphQL/ExtractInputType: Consider moving arguments to a new input type

File: app/models/payment_providers/adyen_provider.rb [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    def merchant_account=(value)
+      push_to_secrets(key: 'merchant_account', value:)
+    end
+
+    def merchant_account
+      get_from_secrets('merchant_account')
+    end
+
+    def live_prefix=(value)
+      push_to_secrets(key: 'live_prefix', value:)
```
Comment (vincent-pochet) [inline app/models/payment_providers/adyen_provider.rb:32]: I don't know if this one is a sensitive , maybe using `push_to_settings` and `get_from_settings` is enough. What do you think?

Maybe it is similar for the `merchant_account` attribute?

File: app/graphql/mutations/payment_providers/adyen.rb (hunk 3 above)
Comment (vincent-pochet) [reply app/graphql/mutations/payment_providers/adyen.rb:13]: This message is a bit "cryptic" :D.

The idea behind this is to define the input arguments in an input type. Here for example, it should be in `Types::PaymentProviders::AdyenInput`. After that, in this mutation, you'll have to replace the arguments with the input type using: 

```ruby
input_object_class Types::PaymentProviders::AdyenInput
```

You can find an example with `Types::Invoices::UpdateInvoiceInput` type, defined in `/app/graphql/types/invoices/update_invoice_input.rb`

File: spec/models/payment_providers/adyen_provider_spec.rb [hunk 19]
```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe PaymentProviders::AdyenProvider, type: :model do
+  let(:provider) { FactoryBot.build_stubbed(:adyen_provider) }
```
Comment (vincent-pochet) [inline spec/models/payment_providers/adyen_provider_spec.rb:6]: To fix the rubocop comments, ou should define this one as spec subject `subject(:provider)` and replace all other ocurrence of `subject` with a dedicated `let`.

**Review by vincent-pochet — COMMENTED** [review]
Looks great! :rocket: 

I've added few comments mainly regarding the rubocop message, but none are blocking.
Let me know if you need more explanation :)

File: app/graphql/mutations/payment_providers/adyen.rb [hunk 20]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+module Mutations
+  module PaymentProviders
+    class Adyen < BaseMutation
```
Comment (rsempe) [inline app/graphql/mutations/payment_providers/adyen.rb:5]: Is it possible to add a spec file for this class?

Something in `spec/graphql/mutations/payment_providers/ayden_spec.rb`

File: app/graphql/types/payment_providers/adyen.rb [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+
+      # NOTE: Api key is a sensitive information. It should not be sent back to the
+      #       front end application. Instead we send an obfuscated value
+      def api_key
+        "#{'•' * 8}…#{object.api_key[-3..]}"
+      end
+
+      def hmac_key
+        "#{'•' * 8}…#{object.hmac_key[-3..]}" if object.hmac_key.present?
+      end
```
Comment (rsempe) [inline app/graphql/types/payment_providers/adyen.rb:22]: Is it possible to add unit tests for these two methods?

File: app/models/payment_providers/adyen_provider.rb (hunk 18 above)
Comment (ivannovosad) [reply app/models/payment_providers/adyen_provider.rb:32]: You're right, I agree, changing it that way - both `live_prefix` and `merchant_account`

File: app/graphql/types/payment_providers/adyen.rb (hunk 21 above)
Comment (ivannovosad) [reply app/graphql/types/payment_providers/adyen.rb:22]: I haven't found specs for graphql types in the project, but those methods are tested in `spec/graphql/mutations/payment_providers/adyen_spec.rb`

**Review by vincent-pochet — APPROVED** [review]

**Pull request merged**